cpp/src/arrow/util/ree_util_test.cc

```diff
@@ -0,0 +1,261 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <gtest/gtest.h>
+
+#include "arrow/array.h"
+#include "arrow/builder.h"
+#include "arrow/compute/api_vector.h"
+#include "arrow/testing/gtest_util.h"
+#include "arrow/type_traits.h"
+#include "arrow/util/logging.h"
+#include "arrow/util/ree_util.h"
+
+namespace arrow {
+namespace ree_util {
+
+template <typename RunEndsType>
+struct ReeUtilTest : public ::testing::Test {
+  // Re-implementation of FindPhysicalIndex that uses trivial linear-search
+  // instead of the more efficient implementation.
+  int64_t FindPhysicalIndexTestImpl(const RunEndsType* run_ends, int64_t run_ends_size,
+                                    int64_t i, int64_t absolute_offset = 0) {
+    for (int64_t j = 0; j < run_ends_size; j++) {
+      if (absolute_offset + i < run_ends[j]) {
+        return j;
+      }
+    }
+    return run_ends_size;
+  }
+};
+TYPED_TEST_SUITE_P(ReeUtilTest);
+
+TYPED_TEST_P(ReeUtilTest, PhysicalOffset) {
+  using RE = TypeParam;  // Run-end type
+  const RE run_ends1[] = {1};
+  ASSERT_EQ(internal::FindPhysicalIndex(run_ends1, 1, 0, 0), 0);
+  const RE run_ends124[] = {1, 2, 4};
+  ASSERT_EQ(internal::FindPhysicalIndex(run_ends124, 3, 0, 0), 0);
+  ASSERT_EQ(internal::FindPhysicalIndex(run_ends124, 3, 1, 0), 1);
+  ASSERT_EQ(internal::FindPhysicalIndex(run_ends124, 3, 2, 0), 2);
+  const RE run_ends234[] = {2, 3, 4};
+  ASSERT_EQ(internal::FindPhysicalIndex(run_ends234, 3, 0, 0), 0);
+  ASSERT_EQ(internal::FindPhysicalIndex(run_ends234, 3, 1, 0), 0);
+  ASSERT_EQ(internal::FindPhysicalIndex(run_ends234, 3, 2, 0), 1);
+  ASSERT_EQ(internal::FindPhysicalIndex(run_ends234, 3, 3, 0), 2);
+  const RE run_ends246[] = {2, 4, 6};
+  ASSERT_EQ(internal::FindPhysicalIndex(run_ends246, 3, 3, 0), 1);
+
+  // Out-of-range logical offset should return run_ends size
```

Comment:
Should we assert that the offset is nonnegative?

Reply:
The returned value here or in `FindPhysicalIndex` implementation?

Reply:
I will add it here and push a commit.

Reply:
I meant in the implementation, sorry

Reply:
I added test cases with negative `i`. I think that's what you meant :)

Reply:
That of course doesn't work. :/ I will add a debug check to `FindPhysicalIndex` instead.